<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting Started with rTASSEL • rTASSEL</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started with rTASSEL">
<meta property="og:description" content="rTASSEL">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rTASSEL</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.17</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/rtassel_walkthrough.html">Getting Started with rTASSEL</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting Started with rTASSEL</h1>
                        <h4 class="author">Brandon Monier</h4>
            
            <h4 class="date">2019-09-04 at 11:21:26</h4>
      
      
      <div class="hidden name"><code>rtassel_walkthrough.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>Thanks for checking out rTASSEL! In this document, we will go over the functionalities used to work with the TASSEL software via R.</p>
<p>TASSEL is a software package used to evaluate traits associations, evolutionary patterns, and linkage disequilibrium. Strengths of this software include:</p>
<ol style="list-style-type: decimal">
<li><p>The opportunity for a number of new and powerful statistical approaches to association mapping such as a General Linear Model (GLM) and Mixed Linear Model (MLM). MLM is an implementation of the technique which our lab’s published Nature Genetics paper - Unified Mixed-Model Method for Association Mapping - which reduces Type I error in association mapping with complex pedigrees, families, founding effects and population structure.</p></li>
<li><p>An ability to handle a wide range of indels (insertion &amp; deletions). Most software ignore this type of polymorphism; however, in some species (like maize), this is the most common type of polymorphism.</p></li>
</ol>
<p>More information can be found in the following paper:</p>
<p>Bradbury PJ, Zhang Z, Kroon DE, Casstevens TM, Ramdoss Y, Buckler ES. (2007) <em>TASSEL: Software for association mapping of complex traits in diverse samples.</em> Bioinformatics 23:2633-2635.</p>
<p>Detailed documentation and source code can be found on our website:</p>
<p><a href="https://www.maizegenetics.net/tassel" class="uri">https://www.maizegenetics.net/tassel</a></p>
</div>
<div id="motivation" class="section level2">
<h2 class="hasAnchor">
<a href="#motivation" class="anchor"></a>Motivation</h2>
<p>The main goal of developing this package is to construct an R-based front-end to connect to a variety of highly used TASSEL methods and analytical tools. By using R as a front-end, we aim to utilize a <em>unified</em> scripting workflow that exploits the analytical prowess of TASSEL in conjunction with R’s popular data handling and parsing capabilities without ever having the user to switch between these two environments.</p>
</div>
<div id="disclaimer" class="section level2">
<h2 class="hasAnchor">
<a href="#disclaimer" class="anchor"></a>Disclaimer</h2>
<p>Due to the <em>experimental</em> nature of this package’s lifecycle, end functionalities are prone to change after end-user input is obtained in the near future.</p>
</div>
</div>
<div id="installation-and-preliminary-steps" class="section level1">
<h1 class="hasAnchor">
<a href="#installation-and-preliminary-steps" class="anchor"></a>Installation and Preliminary Steps</h1>
<div id="prerequisites---installing-rjava" class="section level2">
<h2 class="hasAnchor">
<a href="#prerequisites---installing-rjava" class="anchor"></a>Prerequisites - installing rJava</h2>
<p>Since TASSEL is written in Java, Java JDK will need to be installed on your machine. Additionally, for R to communicate with Java, the R package <code>rJava</code> will need to be installed. Detailed information can be found using the following links, depending on your OS:</p>
<ul>
<li><a href="https://datawookie.netlify.com/blog/2018/02/installing-rjava-on-ubuntu/">Linux</a></li>
<li><a href="https://zhiyzuo.github.io/installation-rJava/">macOS</a></li>
<li><a href="https://cimentadaj.github.io/blog/2018-05-25-installing-rjava-on-windows-10/installing-rjava-on-windows-10/">Windows</a></li>
</ul>
</div>
<div id="problems-with-rjava-if-you-have-upgraded-java" class="section level2">
<h2 class="hasAnchor">
<a href="#problems-with-rjava-if-you-have-upgraded-java" class="anchor"></a>Problems with rJava if you have upgraded Java</h2>
<p>On a MAC: if you previously had <code>rJava</code> working through RStudio, then you upgraded your Java and it now longer works, try the following:</p>
<p>At the command line type:</p>
<pre><code>R CMD javareconf</code></pre>
<p>Then check for a left over symbolic link via:</p>
<pre><code>ls -ltr /usr/local/lib/libjvm.dylib</code></pre>
<p>If the link exists, remove it, then create it fresh via these commands:</p>
<pre><code>rm /usr/local/lib/libjvm.dylib
sudo ln -s $(/usr/libexec/java_home)/lib/server/libjvm.dylib /usr/local/lib</code></pre>
<p>You should now be able to enter RStudio and setup rJava.</p>
</div>
<div id="install-from-bitbucket" class="section level2">
<h2 class="hasAnchor">
<a href="#install-from-bitbucket" class="anchor"></a>Install from BitBucket</h2>
<p>After you have <code>rJava</code> up and running on your machine, install <code>rTASSEL</code> by installing the source code from our BitBucket repository:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="kw">if</span> (!<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="st">"devtools"</span>)) <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"devtools"</span>)
<span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu">install_bitbucket</span>(
    <span class="kw">repo</span> <span class="kw">=</span> <span class="st">"bucklerlab/rtassel"</span>,
    <span class="kw">ref</span> <span class="kw">=</span> <span class="st">"master"</span>,
    <span class="kw">build_vignettes</span> <span class="kw">=</span> <span class="fl">TRUE</span>
)</pre></body></html></div>
<p>After source code has been compiled, the package can be loaded using:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rTASSEL</span>)</pre></body></html></div>
<p>Or, if you want to use a function without violating your environment you can use <code>rTASSEL::&lt;function&gt;</code>, where <code>&lt;function&gt;</code> is an <code>rTASSEL</code> function.</p>
</div>
<div id="preliminary-steps" class="section level2">
<h2 class="hasAnchor">
<a href="#preliminary-steps" class="anchor"></a>Preliminary steps</h2>
<div id="setting-memory" class="section level3">
<h3 class="hasAnchor">
<a href="#setting-memory" class="anchor"></a>Setting Memory</h3>
<p>Since genome-wide association analyses can use up a lot of computational resources, memory allocation to <code>rTASSEL</code> can be modified. To change the amount of memory, use the base <code><a href="https://rdrr.io/r/base/options.html">options()</a></code> function and modify the following parameter:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="kw">java.parameters</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"-Xmx&lt;memory&gt;"</span>, <span class="st">"-Xms&lt;memory&gt;"</span>))</pre></body></html></div>
<p>Replace <code>&lt;memory&gt;</code> with a specified unit of memory. For example, if I want to allocate a maximum of 6 GB of memory for my operations, I would use the input <code>"-Xmx6g"</code>, where <code>g</code> stands for gigabyte (GB). More information about memory allocation can be found <a href="https://stackoverflow.com/questions/14763079/what-are-the-xms-and-xmx-parameters-when-starting-jvm">here</a>.</p>
<p><strong>NOTE:</strong> Setting Java memory options for <code>rTASSEL</code> and any <code>rJava</code>-related packages <em>needs</em> to be set <em>before</em> loading the <code>rTASSEL</code> package!</p>
</div>
<div id="the-importance-of-logging-your-progress" class="section level3">
<h3 class="hasAnchor">
<a href="#the-importance-of-logging-your-progress" class="anchor"></a>The importance of logging your progress</h3>
<p>Before we begin analyzing data, optional parameters can be set up to make <code>rTASSEL</code> more efficient. To prevent your R console from being overloaded with TASSEL logging information, <strong>it is highly recommended that you start a logging file</strong>. This file will house all of TASSEL’s logging output which is beneficial for debugging and tracking the progress of your analytical workflow. To start a logging file, use the following command:</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/startLogger.html">startLogger</a></span>(<span class="kw">fullPath</span> <span class="kw">=</span> <span class="kw">NULL</span>, <span class="kw">fileName</span> <span class="kw">=</span> <span class="kw">NULL</span>)</pre></body></html></div>
<p>If the <code><a href="../reference/startLogger.html">rTASSEL::startLogger()</a></code> parameters are set to <code>NULL</code>, the logging file will be created in your current working directory. If you are unsure of what your working directory is in R, use the base <code><a href="https://rdrr.io/r/base/getwd.html">getwd()</a></code> command.</p>
<p>Additionally, since this is a general walkthrough, certain intricaces of each function may glossed over. If you would like to study a function in full, refer to the R documentation by using <code>?&lt;function&gt;</code> in the console, where <code>&lt;function&gt;</code> is an <code>rTASSEL</code>-based function.</p>
</div>
</div>
</div>
<div id="reading-data" class="section level1">
<h1 class="hasAnchor">
<a href="#reading-data" class="anchor"></a>Reading Data</h1>
<div id="overview-1" class="section level2">
<h2 class="hasAnchor">
<a href="#overview-1" class="anchor"></a>Overview</h2>
<p>Like TASSEL, <code>rTASSEL</code> will read two main types of data:</p>
<ul>
<li>Genotype data</li>
<li>Phenotype data</li>
</ul>
<p>This data can be read in several different ways. In the following examples, we will demonstrate various ways genotype and phenotype information can be loaded into <code>rTASSEL</code> objects.</p>
</div>
<div id="loading-genotype-data" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-genotype-data" class="anchor"></a>Loading genotype data</h2>
<div id="from-a-path" class="section level3">
<h3 class="hasAnchor">
<a href="#from-a-path" class="anchor"></a>From a path</h3>
<p>Currently, reading in genotype data to <code>rTASSEL</code> is based off of file locations as paths. Genotype/sequencing data can be stored in a variety of formats. <code>rTASSEL</code> can read and store a wide variety of file types:</p>
<ul>
<li>hapmap (HMP)</li>
<li>HDF5 (hierarchical data format version 5)</li>
<li>VCF (variant call format)</li>
<li>Plink</li>
</ul>
<p>To load this genotype data, simply store your file location as a string object in R. For this example, we will load two toy data sets - one being a VCF file and the other being a hapmap file. These data sets can be accessed via the <code>rTASSEL</code> package itself:</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="co"># Load hapmap data</span>
<span class="no">genoPathHMP</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(
    <span class="st">"extdata"</span>,
    <span class="st">"mdp_genotype.hmp.txt"</span>,
    <span class="kw">package</span> <span class="kw">=</span> <span class="st">"rTASSEL"</span>
)
<span class="no">genoPathHMP</span></pre></body></html></div>
<pre><code>## [1] "/home/travis/R/Library/rTASSEL/extdata/mdp_genotype.hmp.txt"</code></pre>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="co"># Load VCF data</span>
<span class="no">genoPathVCF</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(
    <span class="st">"extdata"</span>,
    <span class="st">"maize_chr9_10thin40000.recode.vcf"</span>,
    <span class="kw">package</span> <span class="kw">=</span> <span class="st">"rTASSEL"</span>
)
<span class="no">genoPathVCF</span></pre></body></html></div>
<pre><code>## [1] "/home/travis/R/Library/rTASSEL/extdata/maize_chr9_10thin40000.recode.vcf"</code></pre>
<p>Now that we have the file paths to this data, we can pass this to TASSEL and create a formal <code>TasselGenotypePhenotype</code> class object in R using the following:</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co"># Load in hapmap file</span>
<span class="no">tasGenoHMP</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/readGenotypeTableFromPath.html">readGenotypeTableFromPath</a></span>(
    <span class="kw">path</span> <span class="kw">=</span> <span class="no">genoPathHMP</span>
)

<span class="co"># Load in VCF file</span>
<span class="no">tasGenoVCF</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/readGenotypeTableFromPath.html">readGenotypeTableFromPath</a></span>(
    <span class="kw">path</span> <span class="kw">=</span> <span class="no">genoPathVCF</span>
)</pre></body></html></div>
<p>When we call these objects, a summary of the data will be posted to the R console:</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">tasGenoHMP</span></pre></body></html></div>
<pre><code>## A TasselGenotypePhenotype Dataset
##   Class.............. TasselGenotypePhenotype 
##   Taxa............... 281 
##   Positions.......... 3093 
##   Taxa x Positions... 869133 
## ---
##   Genotype Table..... [x]
##   Phenotype Table.... [ ]</code></pre>
<p>This summary details the number of Taxa (<code>Taxa</code>) and marker positions (<code>Positions</code>) within the data set. Additionally, since we can load both genotype and phenotype information into this object, a helpful check will be displayed to show what is populating the object (<code>[x] or [ ]</code>).</p>
</div>
<div id="additional-information-about-tasselphenotypegenotype-data-sets" class="section level3">
<h3 class="hasAnchor">
<a href="#additional-information-about-tasselphenotypegenotype-data-sets" class="anchor"></a>Additional information about TasselPhenotypeGenotype data sets</h3>
<p>In general, this S4 class data object houses “slot” information relating to TASSEL/Java pointers of the respective data.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">tasGenoHMP</span>)</pre></body></html></div>
<pre><code>## [1] "TasselGenotypePhenotype"
## attr(,"package")
## [1] "rTASSEL"</code></pre>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu">slotNames</span>(<span class="no">tasGenoHMP</span>)</pre></body></html></div>
<pre><code>## [1] "name"            "jTasselObj"      "jTaxaList"       "jPositionList"  
## [5] "jGenotypeTable"  "jPhenotypeTable"</code></pre>
<p>Technically, this object does not contain the full information of the data represented in R space, but merely contains addresses to the memory store of the reference TASSEL object ID. For example, if we wanted to extract the <code>GenotypeTable</code> with the S4 <code>@</code> operator, we would get something that looks like this:</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">tasGenoHMP</span>@<span class="kw">jGenotypeTable</span></pre></body></html></div>
<pre><code>## [1] "Java-Object{net.maizegenetics.dna.snp.CoreGenotypeTable@15c43bd9}"</code></pre>
<p>This entity is a <code>rJava</code> internal identifier. It isn’t until we call downstream <code>rTASSEL</code> functions where we will bring the TASSEL data into the R environment.</p>
</div>
</div>
<div id="loading-phenotype-data" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-phenotype-data" class="anchor"></a>Loading phenotype data</h2>
<div id="from-a-path-1" class="section level3">
<h3 class="hasAnchor">
<a href="#from-a-path-1" class="anchor"></a>From a path</h3>
<p>Similar to reading in genotype data, phenotype data can also be read in via paths. If you already have preconstructed phenotype data in a file, this option will most likely work best for you. One caveat to this is how the data file is constructed in terms of columns and trait data for TASSEL analyses. More information about how these files can be found at this <a href="https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/Load/Load">link</a> under the <em>Numerical Data</em> section.</p>
<p>Loading this type of data is very similar to how genotype data is loaded. here, we will use the <code><a href="../reference/readPhenotypeFromPath.html">rTASSEL::readPhenotypeFromPath()</a></code> function:</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="co"># Read from phenotype path</span>
<span class="no">phenoPath</span>  <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"mdp_traits.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"rTASSEL"</span>)
<span class="no">phenoPath</span></pre></body></html></div>
<pre><code>## [1] "/home/travis/R/Library/rTASSEL/extdata/mdp_traits.txt"</code></pre>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="co"># Load into rTASSEL `TasselGenotypePhenotype` object</span>
<span class="no">tasPheno</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/readPhenotypeFromPath.html">readPhenotypeFromPath</a></span>(
    <span class="kw">path</span> <span class="kw">=</span> <span class="no">phenoPath</span>
)

<span class="co"># Inspect object</span>
<span class="no">tasPheno</span></pre></body></html></div>
<pre><code>## A TasselGenotypePhenotype Dataset
##   Class.............. TasselGenotypePhenotype 
##   Taxa............... 301 
##   Positions.......... NA 
##   Taxa x Positions... NA 
## ---
##   Genotype Table..... [ ]
##   Phenotype Table.... [x]
## ---
##   Traits: Taxa EarHT dpoll EarDia</code></pre>
<p>The object output is very similar to the genotype table output with some minor additions to which traits are displayed in the file.</p>
</div>
<div id="from-an-r-data-frame" class="section level3">
<h3 class="hasAnchor">
<a href="#from-an-r-data-frame" class="anchor"></a>From an R data frame</h3>
<p>In some cases you might want to first modify your phenotype data set in R and then load it into the TASSEL environment. If you wish to choose this route, you will need to use the <code><a href="../reference/readPhenotypeFromDataFrame.html">rTASSEL::readPhenotypeFromDataFrame()</a></code> function along with a couple of parameters. First, we will construct an R data frame and load it with this function:</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="co"># Create phenotype data frame</span>
<span class="no">phenoDF</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="no">phenoPath</span>, <span class="kw">header</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">phenoDF</span>)[<span class="fl">1</span>] <span class="kw">&lt;-</span> <span class="st">"Taxon"</span>

<span class="co"># Inspect first few rows</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">phenoDF</span>)</pre></body></html></div>
<pre><code>##   Taxon EarHT  dpoll     EarDia
## 1   811 59.50 -999.0 -999.00000
## 2 33-16 64.75   64.5 -999.00000
## 3 38-11 92.25   68.5   37.89700
## 4  4226 65.50   59.5   32.21933
## 5  4722 81.13   71.5   32.42100
## 6  A188 27.50   62.0   31.41900</code></pre>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="co"># Load into rTASSEL `TasselGenotypePhenotype` object</span>
<span class="no">tasPhenoDF</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/readPhenotypeFromDataFrame.html">readPhenotypeFromDataFrame</a></span>(
    <span class="kw">phenotypeDF</span> <span class="kw">=</span> <span class="no">phenoDF</span>,
    <span class="kw">taxaID</span> <span class="kw">=</span> <span class="st">"Taxon"</span>,
    <span class="kw">attributeTypes</span> <span class="kw">=</span> <span class="kw">NULL</span>
)

<span class="co"># Inspect new object</span>
<span class="no">tasPhenoDF</span></pre></body></html></div>
<pre><code>## A TasselGenotypePhenotype Dataset
##   Class.............. TasselGenotypePhenotype 
##   Taxa............... 301 
##   Positions.......... NA 
##   Taxa x Positions... NA 
## ---
##   Genotype Table..... [ ]
##   Phenotype Table.... [x]
## ---
##   Traits: Taxa EarHT dpoll EarDia</code></pre>
<p>The <code>phenotypeDF</code> parameter is for the R data frame object. The <code>taxaID</code> parameter is needed to determine which column of your data frame is your TASSEL taxa data. The final parameter (<code>attributeTypes</code>) is optional. If this parameter is set to <code>NULL</code>, all remaining data frame columns will be classified as TASSEL <code>data</code> types. If this is not the case for your data (e.g. if you have covariate or factor data in your experiment), <strong>you will need to specify which columns are what TASSEL data type</strong> (i.e. <code>data</code>, <code>covariate</code>, or <code>factor</code>). This will have to be passed as an R vector of string elements (e.g. <code><a href="https://rdrr.io/r/base/c.html">c("data", "factor", "covariate")</a></code>). <em>Currently</em>, this data type needs to be entered in the same order as they are found in the data frame.</p>
</div>
</div>
</div>
<div id="loading-genotype-and-phenotype-data-simultaneously" class="section level1">
<h1 class="hasAnchor">
<a href="#loading-genotype-and-phenotype-data-simultaneously" class="anchor"></a>Loading genotype and phenotype data simultaneously</h1>
<p>In association studies, we are interested in combining our genotype and phenotype data. To usually run this operation in TASSEL, an intersect combination between the two data sets is needed. To run this in <code>rTASSEL</code>, we can use the <code><a href="../reference/readGenotypePhenotype.html">rTASSEL::readGenotypePhenotype()</a></code> function. The parameter input needed for this function is, of course, a genotype and phenotype object. For genotype input, the following can be used:</p>
<ul>
<li>a path to a genotype file</li>
<li>a prior <code>TasselGenotypePhenotype</code> object</li>
</ul>
<p>For phenotype input, the following can be used:</p>
<ul>
<li>a path to a phenotype data set,</li>
<li>a prior <code>TasselGenotypePhenotype</code> object</li>
<li>an R data frame</li>
</ul>
<p>For example, if we wanted to read the prior <code>TasselGenotypePhenotype</code> genotype and phenotype objects from earlier:</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="no">tasGenoPheno</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/readGenotypePhenotype.html">readGenotypePhenotype</a></span>(
    <span class="kw">genoPathOrObj</span> <span class="kw">=</span> <span class="no">tasGenoHMP</span>,
    <span class="kw">phenoPathDFOrObj</span> <span class="kw">=</span> <span class="no">tasPheno</span>
)
<span class="no">tasGenoPheno</span></pre></body></html></div>
<pre><code>## A TasselGenotypePhenotype Dataset
##   Class.............. TasselGenotypePhenotype 
##   Taxa............... 279 
##   Positions.......... 3093 
##   Taxa x Positions... 862947 
## ---
##   Genotype Table..... [x]
##   Phenotype Table.... [x]
## ---
##   Traits: Taxa EarHT dpoll EarDia</code></pre>
<p>We can also use a combination of the above parameter options (e.g. load a genotype path and a phenotype data frame, etc.). One caveat though, <strong>if you load in a phenotype data frame object with this function, the prior parameters from the</strong> <code><a href="../reference/readPhenotypeFromDataFrame.html">rTASSEL::readPhenotypeFromDataFrame</a></code> <strong>will be needed</strong> (i.e. the <code>taxaID</code> and <code>attributeTypes</code> parameters):</p>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="no">tasGenoPhenoDF</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/readGenotypePhenotype.html">readGenotypePhenotype</a></span>(
    <span class="kw">genoPathOrObj</span> <span class="kw">=</span> <span class="no">genoPathHMP</span>,
    <span class="kw">phenoPathDFOrObj</span> <span class="kw">=</span> <span class="no">phenoDF</span>,
    <span class="kw">taxaID</span> <span class="kw">=</span> <span class="st">"Taxon"</span>,
    <span class="kw">attributeTypes</span> <span class="kw">=</span> <span class="kw">NULL</span>
)
<span class="no">tasGenoPhenoDF</span></pre></body></html></div>
<pre><code>## A TasselGenotypePhenotype Dataset
##   Class.............. TasselGenotypePhenotype 
##   Taxa............... 279 
##   Positions.......... 3093 
##   Taxa x Positions... 862947 
## ---
##   Genotype Table..... [x]
##   Phenotype Table.... [x]
## ---
##   Traits: Taxa EarHT dpoll EarDia</code></pre>
</div>
<div id="extracting-data-from-rtassel-objects" class="section level1">
<h1 class="hasAnchor">
<a href="#extracting-data-from-rtassel-objects" class="anchor"></a>Extracting data from rTASSEL objects</h1>
<div id="get-genotype-data" class="section level2">
<h2 class="hasAnchor">
<a href="#get-genotype-data" class="anchor"></a>Get genotype data</h2>
<p>If you want to bring in genotype data into the R environment, you can use the <code><a href="../reference/getSumExpFromGenotypeTable.html">rTASSEL::getSumExpFromGenotypeTable()</a></code> function. All this function needs is a <code>TasselGenotypePhenotype</code> class object <em>containing a genotype table</em>:</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="no">tasSumExp</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/getSumExpFromGenotypeTable.html">getSumExpFromGenotypeTable</a></span>(
    <span class="kw">tasObj</span> <span class="kw">=</span> <span class="no">tasGenoPheno</span>
)
<span class="no">tasSumExp</span></pre></body></html></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 3093 279 
## metadata(0):
## assays(1): ''
## rownames: NULL
## rowData names(3): tasselIndex refAllele altAllele
## colnames(279): 33-16 38-11 ... WF9 YU796NS
## colData names(2): Sample TasselIndex</code></pre>
<p>As you can see above, the gentoype object is returned as a <code>SummarizedExperiment</code> class R object. More information about these objects can be found <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">here</a>.</p>
<p>From this object, we extract taxa information and their respective TASSEL integer locations:</p>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="kw pkg">SummarizedExperiment</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span>(<span class="no">tasSumExp</span>)</pre></body></html></div>
<pre><code>## DataFrame with 279 rows and 2 columns
##              Sample TasselIndex
##         &lt;character&gt;   &lt;integer&gt;
## 33-16         33-16           0
## 38-11         38-11           1
## 4226           4226           2
## 4722           4722           3
## A188           A188           4
## ...             ...         ...
## W22             W22         274
## W64A           W64A         275
## WD               WD         276
## WF9             WF9         277
## YU796NS     YU796NS         278</code></pre>
<p>We can also extract the allelic marker data using <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">SummarizedExperiment::rowData()</a></code>:</p>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="kw pkg">SummarizedExperiment</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span>(<span class="no">tasSumExp</span>)</pre></body></html></div>
<pre><code>## DataFrame with 3093 rows and 3 columns
##      tasselIndex   refAllele   altAllele
##        &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 1              0           A           C
## 2              1           C           G
## 3              2           G           T
## 4              3           A           T
## 5              4           A           G
## ...          ...         ...         ...
## 3089        3088           A           G
## 3090        3089           C           G
## 3091        3090           C           G
## 3092        3091           A           G
## 3093        3092           C           G</code></pre>
<p>And get marker coordinates using <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment::rowRanges()</a></code></p>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="kw pkg">SummarizedExperiment</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">rowRanges</a></span>(<span class="no">tasSumExp</span>)</pre></body></html></div>
<pre><code>## GRanges object with 3093 ranges and 3 metadata columns:
##          seqnames    ranges strand | tasselIndex   refAllele   altAllele
##             &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;integer&gt; &lt;character&gt; &lt;character&gt;
##      [1]        1    157104      + |           0           A           C
##      [2]        1   1947984      + |           1           C           G
##      [3]        1   2914066      + |           2           G           T
##      [4]        1   2914171      + |           3           A           T
##      [5]        1   2915078      + |           4           A           G
##      ...      ...       ...    ... .         ...         ...         ...
##   [3089]       10 147819958      + |        3088           A           G
##   [3090]       10 148332885      + |        3089           C           G
##   [3091]       10 148488692      + |        3090           C           G
##   [3092]       10 148816538      + |        3091           A           G
##   [3093]       10 148907116      + |        3092           C           G
##   -------
##   seqinfo: 10 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="extract-phenotype-data" class="section level2">
<h2 class="hasAnchor">
<a href="#extract-phenotype-data" class="anchor"></a>Extract phenotype data</h2>
<p>If you want to bring in phenotype data into the R environment, you can use the <code><a href="../reference/getPhenotypeDF.html">rTASSEL::getPhenotypeDF()</a></code> function. All this function needs is a <code>TasselGenotypePhenotype</code> class object <em>containing a phenotype table</em>:</p>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="no">tasExportPhenoDF</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/getPhenotypeDF.html">getPhenotypeDF</a></span>(
    <span class="kw">tasObj</span> <span class="kw">=</span> <span class="no">tasGenoPheno</span>
)
<span class="no">tasExportPhenoDF</span></pre></body></html></div>
<pre><code>## DataFrame with 279 rows and 4 columns
##            Taxa     EarHT     dpoll    EarDia
##     &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## 1         33-16     64.75      64.5       NaN
## 2         38-11     92.25      68.5   37.8970
## 3          4226     65.50      59.5   32.2193
## 4          4722     81.13      71.5   32.4210
## 5          A188     27.50      62.0   31.4190
## ...         ...       ...       ...       ...
## 275         W22      44.0      63.5   36.0410
## 276        W64A      59.5      62.5   35.9820
## 277          WD      48.5      62.5   30.4790
## 278         WF9      52.0      64.5   36.6147
## 279     YU796NS      56.5      64.0   40.7550</code></pre>
<p>As shown above, an R <code>tibble</code>-based data frame is exported with converted data types translated from TASSEL. See the following table what TASSEL data types are tranlated into within the R environment:</p>
<table class="table">
<thead><tr class="header">
<th align="left">TASSEL Data</th>
<th align="left">Converted R Data type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">taxa</td>
<td align="left">character</td>
</tr>
<tr class="even">
<td align="left">data</td>
<td align="left">numeric</td>
</tr>
<tr class="odd">
<td align="left">covariate</td>
<td align="left">numeric</td>
</tr>
<tr class="even">
<td align="left">factor</td>
<td align="left">factor</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="filtering-genotype-data" class="section level1">
<h1 class="hasAnchor">
<a href="#filtering-genotype-data" class="anchor"></a>Filtering genotype data</h1>
<p>Prior to association analyses, filtration of genotype data may be necessary. In TASSEL, this accomplished through the Filter menu using two primary plugins:</p>
<ul>
<li>Filter Site Builder plugin</li>
<li>Filter Taxa Builder plugin</li>
</ul>
<p>In <code>rTASSEL</code>, this can also be accomplished using the follwing functions:</p>
<ul>
<li><code><a href="../reference/filterGenotypeTableSites.html">rTASSEL::filterGenotypeTableSites()</a></code></li>
<li><code><a href="../reference/filterGenotypeTableTaxa.html">rTASSEL::filterGenotypeTableTaxa()</a></code></li>
</ul>
<p>These objects take a <code>TasselGenotypePhenotype</code> class object. For example, in our genotype data set, if we want to remove monomorphic and low coverage sites, we could use the following parameters in <code><a href="../reference/filterGenotypeTableSites.html">rTASSEL::filterGenotypeTableSites()</a></code>:</p>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="no">tasGenoPhenoFilt</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/filterGenotypeTableSites.html">filterGenotypeTableSites</a></span>(
    <span class="kw">tasObj</span> <span class="kw">=</span> <span class="no">tasGenoPheno</span>,
    <span class="kw">siteMinCount</span> <span class="kw">=</span> <span class="fl">150</span>,
    <span class="kw">siteMinAlleleFreq</span> <span class="kw">=</span> <span class="fl">0.05</span>,
    <span class="kw">siteMaxAlleleFreq</span> <span class="kw">=</span> <span class="fl">1.0</span>,
    <span class="kw">siteRangeFilterType</span> <span class="kw">=</span> <span class="st">"none"</span>
)
<span class="no">tasGenoPhenoFilt</span></pre></body></html></div>
<pre><code>## A TasselGenotypePhenotype Dataset
##   Class.............. TasselGenotypePhenotype 
##   Taxa............... 279 
##   Positions.......... 2555 
##   Taxa x Positions... 712845 
## ---
##   Genotype Table..... [x]
##   Phenotype Table.... [x]
## ---
##   Traits: Taxa EarHT dpoll EarDia</code></pre>
<p>We can then compare this to our original pre-filtered data set:</p>
<div class="sourceCode" id="cb45"><html><body><pre class="r"><span class="no">tasGenoPheno</span></pre></body></html></div>
<pre><code>## A TasselGenotypePhenotype Dataset
##   Class.............. TasselGenotypePhenotype 
##   Taxa............... 279 
##   Positions.......... 3093 
##   Taxa x Positions... 862947 
## ---
##   Genotype Table..... [x]
##   Phenotype Table.... [x]
## ---
##   Traits: Taxa EarHT dpoll EarDia</code></pre>
<p>These functions can work on any <code>TasselGenotypePhenotype</code> class object that contains genotypic data, regardless of single or combined TASSEL objects.</p>
</div>
<div id="analysis---relatedness" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis---relatedness" class="anchor"></a>Analysis - Relatedness</h1>
<div id="create-a-kinship-matrix-object" class="section level2">
<h2 class="hasAnchor">
<a href="#create-a-kinship-matrix-object" class="anchor"></a>Create a kinship matrix object</h2>
<p>In TASSEL, for mixed linear model analyses, a kinship matrix calculated from genotype data is necessary. This can be accomplished by calculating a kinship TASSEL object using the function <code><a href="../reference/kinshipMatrix.html">rTASSEL::kinshipMatrix()</a></code>. The main parameter input is a <code>TasselGenotypePhenotype</code> class object that contains a genotype data set:</p>
<div class="sourceCode" id="cb47"><html><body><pre class="r"><span class="no">tasKin</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/kinshipMatrix.html">kinshipMatrix</a></span>(<span class="kw">tasObj</span> <span class="kw">=</span> <span class="no">tasGenoPheno</span>)</pre></body></html></div>
<p>Since this object is essentially a TASSEL reference object, it is relatively not of use until association analysis. We can coerce this TASSEL object by using the following function:</p>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="co"># Get full R matrix</span>
<span class="no">tasKinRMat</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/kinshipToRMatrix.html">kinshipToRMatrix</a></span>(<span class="no">tasKin</span>)

<span class="co"># Inspect the first 5 rows and columns</span>
<span class="no">tasKinRMat</span>[<span class="fl">1</span>:<span class="fl">5</span>, <span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##              33-16        38-11        4226         4722        A188
## 33-16  1.796955100  0.039558496  0.06060008 -0.008674403  0.03524417
## 38-11  0.039558496  1.910150800  0.01960260 -0.006317467 -0.02937613
## 4226   0.060600080  0.019602600  1.92645620 -0.017025312  0.02449092
## 4722  -0.008674403 -0.006317467 -0.01702531  1.453315500 -0.02040120
## A188   0.035244170 -0.029376134  0.02449092 -0.020401197  2.00156020</code></pre>
<p>As you can see, we can take the initial TASSEL object and convert it into a standard R object of <code>matrix</code> class.</p>
</div>
<div id="calculate-a-distance-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#calculate-a-distance-matrix" class="anchor"></a>Calculate a distance matrix</h2>
<p>Very similar to kinship matrix calculation, a distance matrix can also be calculated using genotype data</p>
<div class="sourceCode" id="cb50"><html><body><pre class="r"><span class="no">tasDist</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/distanceMatrix.html">distanceMatrix</a></span>(<span class="kw">tasObj</span> <span class="kw">=</span> <span class="no">tasGenoPheno</span>)</pre></body></html></div>
<p>And just like the kinship matrix, we can also coerce this into an R matrix</p>
<div class="sourceCode" id="cb51"><html><body><pre class="r"><span class="co"># Get full R matrix</span>
<span class="no">tasDistRMat</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/distanceToRMatrix.html">distanceToRMatrix</a></span>(<span class="no">tasDist</span>)

<span class="co"># Inspect the first 5 rows and columns</span>
<span class="no">tasDistRMat</span>[<span class="fl">1</span>:<span class="fl">5</span>, <span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##           33-16     38-11      4226      4722      A188
## 33-16 0.0000000 0.2798986 0.2793967 0.2697897 0.2881790
## 38-11 0.2798986 0.0000000 0.2920175 0.2776528 0.3042310
## 4226  0.2793967 0.2920175 0.0000000 0.2863394 0.2999318
## 4722  0.2697897 0.2776528 0.2863394 0.0000000 0.2920304
## A188  0.2881790 0.3042310 0.2999318 0.2920304 0.0000000</code></pre>
</div>
</div>
<div id="analysis---association" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis---association" class="anchor"></a>Analysis - Association</h1>
<div id="overview-2" class="section level2">
<h2 class="hasAnchor">
<a href="#overview-2" class="anchor"></a>Overview</h2>
<p>One of TASSEL’s most powerful functionalities is its capability of performing a variety of different association modeling techniques. If you have started reading the walkthrough here it is <em>strongly suggested that you read the other components of this walkthrough since the following parameters require what we have previously created!</em></p>
<p>If you are not familar with these methods, more information about how these operate in base TASSEL can be found at following links:</p>
<ul>
<li><a href="https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/GLM/GLM">BLUE/GLM</a></li>
<li><a href="https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/MLM/MLM">MLM</a></li>
<li><a href="https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/FastAssociation/FastAssociation">Fast Association</a></li>
</ul>
<p>The <code><a href="../reference/assocModelFitter.html">rTASSEL::assocModelFitter()</a></code> function has several primary components:</p>
<ul>
<li>
<code>tasObj</code>: a <code>TasselGenotypePhenotype</code> class R object</li>
<li>
<code>formula</code>: an R-based linear model formula</li>
<li>
<code>fitMarkers</code>: a boolean parameter to differentiate between BLUE and GLM analyses</li>
<li>
<code>kinship</code>: a TASSEL kinship object</li>
<li>
<code>fastAssociation</code>: a boolean parameter for data sets that have many traits</li>
</ul>
<p>Probably the most important concept of this function is <code>formula</code> parameter. If you are familar with standard R linear model functions, this concept is fairly similar. In TASSEL, a linear model is composed of the following scheme:</p>
<pre><code>y ~ A</code></pre>
<p>…where <code>y</code> is any TASSEL <code>data</code> type and <code>A</code> is any TASSEL <code>covariate</code> and / or <code>factor</code> types:</p>
<pre><code>&lt;data&gt; ~ &lt;covariate&gt; and/or &lt;factor&gt;</code></pre>
<p>This model can be written out in several ways. With an example phenotype data, we can have the following variables that are represented in TASSEL in the following way:</p>
<ul>
<li>
<code>Taxon</code> <code>&lt;taxa&gt;</code>
</li>
<li>
<code>EarHT</code> <code>&lt;data&gt;</code>
</li>
<li>
<code>dpoll</code> <code>&lt;data&gt;</code>
</li>
<li>
<code>EarDia</code> <code>&lt;data&gt;</code>
</li>
<li>
<code>location</code> <code>&lt;factor&gt;</code>
</li>
<li>
<code>Q1</code> <code>&lt;covariate&gt;</code>
</li>
<li>
<code>Q2</code> <code>&lt;covariate&gt;</code>
</li>
<li>
<code>Q3</code> <code>&lt;covariate&gt;</code>
</li>
</ul>
<p>Using this data, we could write out the following formula in R</p>
<pre><code>list(EarHT, dpoll, EarDia) ~ location + Q1 + Q2 + Q3</code></pre>
<p>In the above example, we use a base <code><a href="https://rdrr.io/r/base/list.html">list()</a></code> function to indicate analysis on multiple numeric data types. For covariate and factor information, we use <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code> operator. One problem with this implementation is that it can become cumbersome and prone to error if we want to analyze the entirety of a large data set or all data and/or factor and covariate types.</p>
<p>A work around for this problem is to utilize a special character to indicate all elements within the model (<code>.</code>). By using the <code>.</code> operator we can simplify the above model into the following:</p>
<pre><code>. ~ .</code></pre>
<p>This indicates we want to analyze <em>the whole</em> data set and leave nothing out. If we want to analyze all data types and only a handful of factor and/or covariates, we can use something like this:</p>
<pre><code>. ~ location + Q1 + Q2</code></pre>
<p>Or vice-versa:</p>
<pre><code>list(EarHT, dpoll) ~ .</code></pre>
<p>Take note we can be very specific with what we want to include in our trait model! In the above example we have deliberately left out <code>EarDia</code> from our model.</p>
<p>Additionally, we can also fit marker and kinship data to our model which can change our analytical methods. Since these options in TASSEL are binary, additional parameters are passed for this function. In this case, genotype/marker data is fitted using the <code>fitMarker</code> parameter and kinship is fitted using the <code>kinship</code> parameter.</p>
<p>Fast Association implements methods described by Shabalin (2012). This method provides an ordinary least squares solution for fixed effect models. For this method to proper work it is necessary that your have:</p>
<ul>
<li><em>No missing data in your phenotype data set</em></li>
<li>
<em>Phenotypes and genotypes have been merged using an intersect join</em>. Since this is currently the only option of join genotype and phenotype data, you do not have to worry about this <em>for now</em>.</li>
</ul>
<p><strong>NOTE</strong>: since we are working with “toy” data, empirical insight will not be elucidated upon in the following steps. This is simply to show the user how properly use these functions and the outputs that they give.</p>
<p>In the following examples, we will run example data and in return, obtain TASSEL association table reports in the form of an R <code>list</code> object containing <code>tibble</code>-based R data frames.</p>
</div>
<div id="calculate-blues" class="section level2">
<h2 class="hasAnchor">
<a href="#calculate-blues" class="anchor"></a>Calculate BLUEs</h2>
<p>To caclulate best linear unbiased estimates (BLUEs), numeric phenotype data can be used along with covariate and factor data <em>only</em> if it is intended to control for field variation. Since genotype data is not needed for this method, we can leave the <code>fitMarkers</code>, <code>kinship</code>, and <code>fastAssociation</code> to <code>NULL</code> or <code>FALSE</code>:</p>
<div class="sourceCode" id="cb59"><html><body><pre class="r"><span class="co"># Read in phenotype data</span>
<span class="no">phenoPathCov</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"mdp_phenotype.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"rTASSEL"</span>)
<span class="no">tasPhenoCov</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/readPhenotypeFromPath.html">readPhenotypeFromPath</a></span>(<span class="no">phenoPathCov</span>)

<span class="co"># Calculate BLUEs</span>
<span class="no">tasBLUE</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/assocModelFitter.html">assocModelFitter</a></span>(
    <span class="kw">tasObj</span> <span class="kw">=</span> <span class="no">tasPhenoCov</span>,
    <span class="kw">formula</span> <span class="kw">=</span> <span class="no">.</span> ~ <span class="no">.</span>,                  <span class="co"># &lt;- All data is used!</span>
    <span class="kw">fitMarkers</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
    <span class="kw">kinship</span> <span class="kw">=</span> <span class="kw">NULL</span>,
    <span class="kw">fastAssociation</span> <span class="kw">=</span> <span class="fl">FALSE</span>
)

<span class="co"># Return BLUE output</span>
<span class="no">tasBLUE</span></pre></body></html></div>
</div>
<div id="calculate-glm" class="section level2">
<h2 class="hasAnchor">
<a href="#calculate-glm" class="anchor"></a>Calculate GLM</h2>
<p>Similar to BLUEs, we can fit a generalized linear model (GLM) by simply fitting marker data to our model. For this, we need a genotype data set combined with our phenotype data in a <code>TasselGenotypePhenotype</code> class object:</p>
<div class="sourceCode" id="cb60"><html><body><pre class="r"><span class="co"># Calculate GLM</span>
<span class="no">tasGLM</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/assocModelFitter.html">assocModelFitter</a></span>(
    <span class="kw">tasObj</span> <span class="kw">=</span> <span class="no">tasGenoPheno</span>,             <span class="co"># &lt;- our prior TASSEL object</span>
    <span class="kw">formula</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">EarHT</span>, <span class="no">dpoll</span>) ~ <span class="no">.</span>,  <span class="co"># &lt;- only EarHT and dpoll are ran</span>
    <span class="kw">fitMarkers</span> <span class="kw">=</span> <span class="fl">TRUE</span>,                 <span class="co"># &lt;- set this to TRUE for GLM</span>
    <span class="kw">kinship</span> <span class="kw">=</span> <span class="kw">NULL</span>,
    <span class="kw">fastAssociation</span> <span class="kw">=</span> <span class="fl">FALSE</span>
)

<span class="co"># Return GLM output</span>
<span class="no">tasGLM</span></pre></body></html></div>
</div>
<div id="calculate-mlm" class="section level2">
<h2 class="hasAnchor">
<a href="#calculate-mlm" class="anchor"></a>Calculate MLM</h2>
<p>Adding to our complexity, we can fit a mixed linear model (MLM) by adding kinship to our analysis. In addition to the prior parameters, we will also need a TASSEL kinship object (see <em>Create a kinship matrix object</em> in the <em>Analysis - Relatedness</em> section):</p>
<div class="sourceCode" id="cb61"><html><body><pre class="r"><span class="co"># Calculate MLM</span>
<span class="no">tasMLM</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/assocModelFitter.html">assocModelFitter</a></span>(
    <span class="kw">tasObj</span> <span class="kw">=</span> <span class="no">tasGenoPheno</span>,             <span class="co"># &lt;- our prior TASSEL object</span>
    <span class="kw">formula</span> <span class="kw">=</span> <span class="no">EarHT</span> ~ <span class="no">.</span>,               <span class="co"># &lt;- run only EarHT</span>
    <span class="kw">fitMarkers</span> <span class="kw">=</span> <span class="fl">TRUE</span>,                 <span class="co"># &lt;- set this to TRUE for GLM</span>
    <span class="kw">kinship</span> <span class="kw">=</span> <span class="no">tasKin</span>,                  <span class="co"># &lt;- our prior kinship object</span>
    <span class="kw">fastAssociation</span> <span class="kw">=</span> <span class="fl">FALSE</span>
)

<span class="co"># Return GLM output</span>
<span class="no">tasMLM</span></pre></body></html></div>
</div>
<div id="calculate-fast-association" class="section level2">
<h2 class="hasAnchor">
<a href="#calculate-fast-association" class="anchor"></a>Calculate Fast Association</h2>
<p>Finally, we can run fast association analysis in our GLM model by setting the <code>fastAssociation</code> parameter to <code>TRUE</code>. <strong>NOTE</strong>: this is only really effective if you have many phenotype traits:</p>
<div class="sourceCode" id="cb62"><html><body><pre class="r"><span class="co"># Read data - need only non missing data!</span>
<span class="no">phenoPathFast</span> <span class="kw">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(
    <span class="st">"extdata"</span>,
    <span class="st">"mdp_traits_nomissing.txt"</span>,
    <span class="kw">package</span> <span class="kw">=</span> <span class="st">"rTASSEL"</span>
)

<span class="co"># Creat rTASSEL object - use prior TASSEL genotype object</span>
<span class="no">tasGenoPhenoFast</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/readGenotypePhenotype.html">readGenotypePhenotype</a></span>(
    <span class="kw">genoPathOrObj</span> <span class="kw">=</span> <span class="no">tasGenoHMP</span>,
    <span class="kw">phenoPathDFOrObj</span> <span class="kw">=</span> <span class="no">phenoPathFast</span>
)


<span class="co"># Calculate MLM</span>
<span class="no">tasFAST</span> <span class="kw">&lt;-</span> <span class="kw pkg">rTASSEL</span><span class="kw ns">::</span><span class="fu"><a href="../reference/assocModelFitter.html">assocModelFitter</a></span>(
    <span class="kw">tasObj</span> <span class="kw">=</span> <span class="no">tasGenoPhenoFast</span>,         <span class="co"># &lt;- our prior TASSEL object</span>
    <span class="kw">formula</span> <span class="kw">=</span> <span class="no">.</span> ~ <span class="no">.</span>,                   <span class="co"># &lt;- run all of the phenotype data</span>
    <span class="kw">fitMarkers</span> <span class="kw">=</span> <span class="fl">TRUE</span>,                 <span class="co"># &lt;- set this to TRUE for GLM</span>
    <span class="kw">kinship</span> <span class="kw">=</span> <span class="kw">NULL</span>,
    <span class="kw">fastAssociation</span> <span class="kw">=</span> <span class="fl">TRUE</span>             <span class="co"># &lt;- set this to TRUE for fast assoc.</span>
)

<span class="co"># Return GLM output</span>
<span class="no">tasFAST</span></pre></body></html></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Brandon Monier, Terry Casstevens, Ed Buckler.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
